### PR TITLE
Worked on the implementation of Generalized Wasserstein Dice Loss

### PR DESCRIPTION
**Pull request by arp95**
Worked on issue #601 . @wyli Please check.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
There are some minor code format issues:
```
/home/runner/work/MONAI/MONAI/monai/losses/dice.py:342:166: B950 line too long (165 > 120 characters)
/home/runner/work/MONAI/MONAI/monai/losses/dice.py:407:10: N806 variable 'M_extended' in function should be lowercase
/home/runner/work/MONAI/MONAI/monai/losses/dice.py:408:10: N806 variable 'M_extended' in function should be lowercase
/home/runner/work/MONAI/MONAI/monai/losses/dice.py:409:10: N806 variable 'M_extended' in function should be lowercase
/home/runner/work/MONAI/MONAI/monai/losses/dice.py:419:10: N806 variable 'M_extended' in function should be lowercase
/home/runner/work/MONAI/MONAI/monai/losses/dice.py:420:10: N806 variable 'M_extended' in function should be lowercase
```
Could you please help update it?
Thanks.

**Comment by arp95** [comment]
Sure. will look into it.

**Comment by arp95** [comment]
@Nic-Ma Resolved the issues. Please check.

**Comment by Nic-Ma** [comment]
Thanks, there is still some type issue:
```
monai/losses/dice.py:374: error: Module has no attribute "LongTensor"
monai/losses/dice.py:439: error: Argument 1 to "unsqueeze" has incompatible type "float"; expected "Tensor"
monai/losses/dice.py:456: error: Argument 1 to "unsqueeze" has incompatible type "float"; expected "Tensor"
```

**Comment by arp95** [comment]
@Nic-Ma That was because I was converting the target to LongTensor type. But I dont think we need that as for other loss functions we havent converted target to LongTensor type.
Updated, please check.

File: monai/losses/dice.py [hunk 1]
```diff
@@ -298,5 +300,177 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
...
+            flat_target: the target tensor.
+        """
+        one_hot = F.one_hot(flat_target, num_classes=self.num_classes).permute(0, 2, 1).float()
+        volumes = torch.sum(one_hot, dim=2)
+        alpha = 1.0 / (volumes + 1.0)
+        return alpha
+
+
 dice = Dice = DiceLoss
 generalized_dice = GeneralizedDiceLoss
```
Comment (ericspod) [inline monai/losses/dice.py:476]: We should add an alias here for `GeneralizedWassersteinDiceLoss`, so something like `wasserstein_dice = GeneralizedWassersteinDiceLoss`.

File: monai/losses/dice.py [hunk 2]
```diff
@@ -298,5 +300,177 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
...
+        super(GeneralizedWassersteinDiceLoss, self).__init__(reduction=reduction)
+
+        if dist_matrix.shape[0] != dist_matrix.shape[1]:
+            raise ValueError("Dist Matrix is invalid.")
+
+        self.M = dist_matrix
+        if isinstance(self.M, np.ndarray):
+            self.M = torch.from_numpy(self.M)
+        if torch.cuda.is_available():
+            self.M = self.M.cuda()
```
Comment (ericspod) [inline monai/losses/dice.py:360]: This will move M to device 0 which won't necessarily be compatible with multi-GPU code. What I think we have to do is move M to the correct device in `forward`, in which case you would have a local copy of M constructed by `M = torch.clone(self.M).to(input.get_device())` which is used in place of `self.M` everywhere. I think this is the correct way of doing it although it involves copying M every forward pass, alternatively you could recall what the previous value of `input.get_device()` was and reuse the created M copy if it's same.

**Comment by ericspod** [comment]
We do also need unit tests for this type, we can do simple ones that just copy those for the existing `DiceLoss` type.

File: monai/losses/dice.py (hunk 1 above)
Comment (arp95) [reply monai/losses/dice.py:476]: yes adding the alias.

File: monai/losses/dice.py (hunk 2 above)
Comment (arp95) [reply monai/losses/dice.py:360]: thats something i wasnt aware of. i will update the code with your suggestion.

**Comment by arp95** [comment]
@Nic-Ma Added the test class for the Generalized Wasserstein Dice Loss. Also added the citation of the code in the file. Please check.

**Comment by arp95** [comment]
@Nic-Ma Can you please detail the re-formatting error I am getting. Rest test cases are running ok.

File: monai/losses/dice.py [hunk 3]
```diff
@@ -298,5 +300,177 @@ def forward(self, input: torch.Tensor, target: torch.Tensor, smooth: float = 1e-
...
+        if dist_matrix.shape[0] != dist_matrix.shape[1]:
+            raise ValueError("Dist Matrix is invalid.")
+
+        self.M = dist_matrix
+        if isinstance(self.M, np.ndarray):
+            self.M = torch.from_numpy(self.M)
+        if torch.max(self.M) != 1:
+            self.M = self.M / torch.max(self.M)
+        self.num_classes = self.M.size(0)
+        self.reduction = reduction
```
Comment (wyli) [inline monai/losses/dice.py:362]: `self.reduction = reduction` is done in the parent class already, I think we can remove line 362

Comment (arp95) [reply monai/losses/dice.py:362]: I am not using DiceLoss as the parent class here. I am using _Loss from torch.

Comment (wyli) [reply monai/losses/dice.py:362]: _Loss also has a reduction variable, could follow this https://github.com/Project-MONAI/MONAI/blob/67741bfd8b6809be299658acd787f01f520cbd6a/monai/losses/dice.py#L69

Comment (arp95) [reply monai/losses/dice.py:362]: @wyli thanks will look into this. also how do I properly cite the resource I am referring to? Like the tests are failing because I am referring to the citation from where I referred the code.

# @inproceedings{fidon2017generalised,
#  title={Generalised {W}asserstein dice score for imbalanced multi-class 
#         segmentation using holistic convolutional networks},
#  author={Fidon, Lucas and Li, Wenqi and Garcia-Peraza-Herrera, Luis C 
#          and Ekanayake, Jinendra and Kitchen, Neil and Ourselin, 
#          S{\'e}bastien and #Vercauteren, Tom},
#  booktitle={International MICCAI Brainlesion Workshop},
#  pages={64--76},
#  year={2017},
#  organization={Springer}
# }

Comment (arp95) [reply monai/losses/dice.py:362]: @wyli thanks will look into this. also how do I properly cite the resource I am referring to? Like the tests are failing because I am referring to the citation from where I referred the code.

"""
@inproceedings{fidon2017generalised,
title={Generalised {W}asserstein dice score for imbalanced multi-class segmentation using holistic convolutional networks},
author={Fidon, Lucas and Li, Wenqi and Garcia-Peraza-Herrera, Luis C and Ekanayake, Jinendra and Kitchen, Neil and Ourselin, S{\'e}bastien and #Vercauteren, Tom},
booktitle={International MICCAI Brainlesion Workshop},pages={64--76}, year={2017},
organization={Springer}
"""

Comment (wyli) [reply monai/losses/dice.py:362]: those test cases are not from the publication, so you can just remove the paragraph. the `GeneralizedWassersteinDiceLoss` you've created already nicely covers all the reference information

Comment (arp95) [reply monai/losses/dice.py:362]: ok thanks. working on it.

**Comment by arp95** [comment]
@Nic-Ma @wyli Please check now.
